File: engine/tokenize.py
"""تنظيف الشطر وبناء قائمة الصوامت مع نوع كل حرف."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Literal, Optional

from .letters import (
    ALIF_MADD,
    ALLAH_FORMS,
    FUNCTION_10,
    HARAKAT,
    LAM,
    MADDA_ALIF,
    PREFIX_CONNECT,
    PRONOUNS,
    SHADDA,
    SHAMSI,
    SUKUN,
    TATWEEL,
    TA_MARBUTA,
    VOWEL_MARKS,
    WAW,
    YA,
    is_arabic_letter,
    strip_harakat,
)

Kind = Literal[
    "cons",
    "alif_madd",
    "waw",
    "ya",
    "wasl",
    "ta_marbuta",
    "collapsed",
]


@dataclass
class Phoneme:
    """وحدة تقطيع واحدة = حرف منطوق يحتمل 1 أو 0."""

    char: str
    kind: Kind
    fixed: Optional[int]  # 0 / 1 / None
    display: bool
    word_i: int
    src_index: int  # موقع في النص المنظّف
    hint: str = ""
    lockable: bool = True


@dataclass
class Hemistich:
    raw: str
    cleaned: str
    words: list[str]
    phonemes: list[Phoneme] = field(default_factory=list)


_KEEP_RE = re.compile(r"[^\u0600-\u06FF\s\*]+")
_MULTI_SPACE = re.compile(r"\s+")
_NUMBERING = re.compile(r"^[\d٠-٩]+[.\-)\]]\s*")


def split_bayt(text: str) -> list[str]:
    """قسم البيت إلى صدر وعجز إذا وجد *** أو * أو سطر جديد."""
    if not text or not text.strip():
        return []
    t = text.replace("\r\n", "\n").replace("\r", "\n")
    if "***" in t:
        parts = [p.strip() for p in t.split("***")]
    elif "\n" in t:
        parts = [p.strip() for p in t.split("\n")]
    elif " * " in t:
        parts = [p.strip() for p in t.split(" * ")]
    elif t.count("*") == 1:
        parts = [p.strip() for p in t.split("*")]
    else:
        parts = [t.strip()]
    return [p for p in parts if p]


def clean_line(text: str) -> str:
    t = text.strip()
    t = t.replace(TATWEEL, "")
    t = _NUMBERING.sub("", t)
    t = _KEEP_RE.sub(" ", t)
    t = t.replace("*", " ")
    t = _MULTI_SPACE.sub(" ", t).strip()
    return t


def _iter_base_letters(word: str):
    """أخرج (حرف، شدة؟، سكون مكتوب؟، حركة مكتوبة؟) مع فك الشدة فوراً."""
    i = 0
    n = len(word)
    while i < n:
        ch = word[i]
        if ch in HARAKAT or ch == TATWEEL or ch == " ":
            i += 1
            continue
        if not is_arabic_letter(ch) and ch not in "آأإؤئءىةٱ":
            i += 1
            continue
        j = i + 1
        has_shadda = False
        has_sukun = False
        has_vowel = False
        while j < n and word[j] in HARAKAT:
            if word[j] == SHADDA:
                has_shadda = True
            elif word[j] == SUKUN:
                has_sukun = True
            elif word[j] in VOWEL_MARKS:
                has_vowel = True
            j += 1
        if has_shadda:
            # الحرف المشدد: ساكن ثم متحرك
            yield ch, False, True, False
            yield ch, False, False, True
        else:
            yield ch, False, has_sukun, has_vowel
        i = j


def _is_allah_word(bare: str) -> bool:
    return bare in ALLAH_FORMS or bare.endswith("الله") or bare == "الله"


def _allah_phonemes(word: str, word_i: int, first_word: bool, src0: int) -> list[Phoneme]:
    """
    كلمة الله ≈ ألالاه: 10 ثم 10. ليست وتداً وحده.
    بالوصل تُحذف ألف الله فيُلصق اللام السابق.
    """
    bare = strip_harakat(word)
    prefix = ""
    core = bare
    for p in ("و", "ب", "ف", "ت", "ل"):
        if bare.startswith(p) and bare[len(p) :] in ("له", "الله") or (
            bare.startswith(p) and "الله" in bare[len(p) :]
        ):
            if bare.startswith(p + "الله") or bare == p + "له":
                prefix = p
                core = bare[len(p) :]
                break
    if core == "له":
        # لله
        prefix = prefix or "ل"
        core = "الله"

    out: list[Phoneme] = []
    if prefix:
        out.append(
            Phoneme(
                char=prefix,
                kind="cons",
                fixed=1,
                display=True,
                word_i=word_i,
                src_index=src0,
                hint="prefix",
            )
        )

    # أ ل ل ا ه
    alif_fixed: Optional[int]
    if first_word and not prefix:
        alif_fixed = 1
        alif_kind: Kind = "cons"
    else:
        # همزة وصل محتملة
        alif_fixed = None
        alif_kind = "wasl"

    out.append(
        Phoneme(
            char="ا",
            kind=alif_kind,
            fixed=alif_fixed if alif_kind != "wasl" else None,
            display=True,
            word_i=word_i,
            src_index=src0,
            hint="allah_alif",
        )
    )
    out.append(
        Phoneme(
            char="ل",
            kind="cons",
            fixed=0,
            display=True,
            word_i=word_i,
            src_index=src0,
            hint="allah_lam1",
        )
    )
    out.append(
        Phoneme(
            char="ل",
            kind="cons",
            fixed=1,
            display=True,
            word_i=word_i,
            src_index=src0,
            hint="allah_lam2",
        )
    )
    out.append(
        Phoneme(
            char="ا",
            kind="alif_madd",
            fixed=0,
            display=True,
            word_i=word_i,
            src_index=src0,
            hint="allah_madd",
        )
    )
    out.append(
        Phoneme(
            char="ه",
            kind="collapsed",
            fixed=0,
            display=True,
            word_i=word_i,
            src_index=src0,
            hint="allah_ha",
        )
    )
    return out


def _function_10(word: str, word_i: int, src0: int) -> list[Phoneme]:
    letters = [ch for ch, *_ in _iter_base_letters(word)]
    if not letters:
        letters = list(strip_harakat(word))
    if len(letters) == 1:
        # لا / ما as V + madd already 2 in يا; single like و handled elsewhere
        ch = letters[0]
        if ch in ALIF_MADD or ch == "ى":
            return [
                Phoneme(ch, "alif_madd", 0, True, word_i, src0, "fn"),
            ]
        return [
            Phoneme(ch, "cons", 1, True, word_i, src0, "fn"),
            Phoneme("", "collapsed", 0, False, word_i, src0, "fn_sukun"),
        ]
    out: list[Phoneme] = []
    # أول متحرك، الباقي ساكن (يا، من، في، لا، ما، قد، هل، بل، لم، إن، أن، أو، عن)
    for k, ch in enumerate(letters):
        if k == 0:
            kind: Kind = "cons"
            if ch in (WAW,):
                kind = "waw"
            elif ch in (YA, "ى"):
                kind = "ya"
            out.append(Phoneme(ch, kind, 1, True, word_i, src0, "fn10"))
        elif ch in ALIF_MADD or ch == "ى" or ch == MADDA_ALIF:
            out.append(Phoneme("ا" if ch == MADDA_ALIF else ch, "alif_madd", 0, True, word_i, src0, "fn10"))
        else:
            out.append(Phoneme(ch, "cons", 0, True, word_i, src0, "fn10"))
    return out


def _pronoun(word: str, word_i: int, src0: int, last_word: bool) -> list[Phoneme]:
    """هو/هي/هم: الواو والياء صامتان لا مدّ. الأول متحرك."""
    letters = [ch for ch, *_ in _iter_base_letters(word)]
    if not letters:
        letters = list(strip_harakat(word))
    out: list[Phoneme] = []
    for k, ch in enumerate(letters):
        last = last_word and k == len(letters) - 1
        fixed: Optional[int] = 0 if last else 1
        out.append(
            Phoneme(
                char=ch,
                kind="cons",
                fixed=fixed,
                display=True,
                word_i=word_i,
                src_index=src0,
                hint="pron",
            )
        )
    return out


def tokenize_hemistich(text: str) -> Hemistich:
    cleaned = clean_line(text)
    words = [w for w in cleaned.split(" ") if w]
    h = Hemistich(raw=text, cleaned=cleaned, words=words)
    if not words:
        return h

    src = 0
    for wi, word in enumerate(words):
        bare = strip_harakat(word)
        first = wi == 0

        if _is_allah_word(bare):
            h.phonemes.extend(_allah_phonemes(word, wi, first, src))
            src += len(word) + 1
            continue

        if bare in FUNCTION_10:
            h.phonemes.extend(_function_10(word, wi, src))
            src += len(word) + 1
            continue

        if bare in PRONOUNS:
            h.phonemes.extend(_pronoun(word, wi, src, last_word=(wi == len(words) - 1)))
            src += len(word) + 1
            continue

        # ال التعريف
        letters = list(_iter_base_letters(word))
        idx = 0
        if (
            len(letters) >= 3
            and letters[0][0] in ALIF_MADD | set("أإٱآ")
            and letters[1][0] == LAM
        ):
            shamsi = letters[2][0] in SHAMSI
            # ألف ال: وصل إن لم تكن أول كلمة
            alif_ch = letters[0][0]
            if alif_ch == MADDA_ALIF:
                h.phonemes.append(Phoneme("آ", "cons", 1, True, wi, src, "hamza"))
                h.phonemes.append(Phoneme("ا", "alif_madd", 0, True, wi, src, "madda"))
            elif first:
                h.phonemes.append(
                    Phoneme("ا", "cons", 1, True, wi, src, "al_alif")
                )
            else:
                h.phonemes.append(
                    Phoneme("ا", "wasl", None, True, wi, src, "al_wasl")
                )
            if shamsi:
                # اللام تُدغم: لا تُحسب مستقلة. شدة الحرف الشمسي 0 ثم 1
                # letters[2] already may have shadda unfolded; force 0 then 1
                sham = letters[2][0]
                h.phonemes.append(
                    Phoneme(sham, "cons", 0, True, wi, src, "shamsi_sukun")
                )
                h.phonemes.append(
                    Phoneme(sham, "cons", 1, False, wi, src, "shamsi_move")
                )
                idx = 3
            else:
                h.phonemes.append(
                    Phoneme("ل", "cons", 0, True, wi, src, "qamari_lam")
                )
                idx = 2

        rest = letters[idx:]
        for k, (ch, _sh, has_sukun, has_vowel) in enumerate(rest):
            last_of_word = k == len(rest) - 1
            last_of_line = last_of_word and wi == len(words) - 1

            if ch == MADDA_ALIF:
                h.phonemes.append(Phoneme("أ", "cons", 1, True, wi, src, "madda_h"))
                h.phonemes.append(Phoneme("ا", "alif_madd", 0, True, wi, src, "madda_a"))
                continue
            if ch in ALIF_MADD:
                h.phonemes.append(Phoneme("ا", "alif_madd", 0, True, wi, src, "madd"))
                continue
            if ch == "ى":
                h.phonemes.append(Phoneme("ى", "alif_madd", 0, True, wi, src, "maqsur"))
                continue
            if ch == TA_MARBUTA:
                # ة في الوقف هاء ساكنة
                h.phonemes.append(
                    Phoneme("ة", "ta_marbuta", 0 if last_of_line else None, True, wi, src, "ta")
                )
                continue

            kind: Kind = "cons"
            if ch == WAW:
                kind = "waw"
            elif ch == YA:
                kind = "ya"

            fixed: Optional[int] = None
            if has_sukun:
                fixed = 0
            elif has_vowel:
                fixed = 1
            if last_of_line:
                fixed = 0

            # واو/ياء وسط الكلمة محتملتان مداً إن لم تُقيَّدا بحركة مكتوبة
            if kind in ("waw", "ya") and not last_of_line and not first and k > 0:
                if has_vowel:
                    fixed = 1
                elif has_sukun:
                    fixed = 0
                else:
                    fixed = None

            # حرف صغير في أول الكلمة: متحرك يُوصل
            if k == 0 and idx == 0 and len(rest) == 1 and ch in PREFIX_CONNECT and not last_of_line:
                fixed = 1

            h.phonemes.append(
                Phoneme(ch, kind, fixed, True, wi, src, kind)
            )

        src += len(word) + 1

    # آخر حرف في الشطر ساكن — قيد ثابت
    if h.phonemes:
        last_disp = None
        for p in reversed(h.phonemes):
            if p.display or p.kind != "collapsed":
                last_disp = p
                break
        if last_disp is not None and last_disp.kind != "wasl":
            last_disp.fixed = 0

    _apply_madd_constraints(h)
    return h


def _apply_madd_constraints(h: Hemistich) -> None:
    """الألف بعد صامت = مد 0، والحرف قبل المد متحرك. المد+سكون يُطوي بتّاً واحداً."""
    ph = h.phonemes
    for i, p in enumerate(ph):
        if p.kind == "alif_madd":
            j = i - 1
            while j >= 0 and ph[j].kind in ("wasl", "collapsed"):
                j -= 1
            if j >= 0 and ph[j].kind not in ("alif_madd",):
                if ph[j].fixed is None:
                    ph[j].fixed = 1
                elif ph[j].fixed == 0 and ph[j].hint not in ("fn10", "qamari_lam", "allah_lam1"):
                    # تعارض نادر: نترك البحث يفشل هذا الفرع
                    pass
        # سكون بعد ألف مد (جال، لاه، تمنى) لا يضيف بتّاً
        if i > 0 and ph[i - 1].kind == "alif_madd" and p.fixed == 0:
            if p.kind in ("cons", "ta_marbuta", "collapsed"):
                p.kind = "collapsed"

File: src/lib/arud/tokenize.ts
import type { Hemistich, Kind, Phoneme } from "./types";

const SHAMSI = new Set([... "تثدذرزسشصضطظلن"]);
const ALIF_MADD = new Set(["ا", "ٱ"]);
const MADDA_ALIF = "آ";
const WAW = "و";
const YA = "ي";
const TA_MARBUTA = "ة";
const LAM = "ل";
const SHADDA = "\u0651";
const SUKUN = "\u0652";
const FATHA = "\u064e";
const DAMMA = "\u064f";
const KASRA = "\u0650";
const TATWEEL = "\u0640";
const SUPERSCRIPT_ALIF = "\u0670";
const FATHATAN = "\u064b";
const DAMMATAN = "\u064c";
const KASRATAN = "\u064d";

const HARAKAT = new Set([
  SHADDA,
  SUKUN,
  FATHA,
  DAMMA,
  KASRA,
  FATHATAN,
  DAMMATAN,
  KASRATAN,
  SUPERSCRIPT_ALIF,
]);
const VOWEL_MARKS = new Set([FATHA, DAMMA, KASRA]);
const PREFIX_CONNECT = new Set([..."وفبكل"]);

const FUNCTION_10 = new Set([
  "يا",
  "من",
  "عن",
  "في",
  "لم",
  "هل",
  "بل",
  "قد",
  "إن",
  "أن",
  "او",
  "أو",
  "ما",
  "لا",
  "له",
  "به",
  "كم",
  "ثم",
]);

const PRONOUNS = new Set(["هو", "هي", "هم", "هن", "هما"]);

const ALLAH_FORMS = new Set([
  "الله",
  "اللّه",
  "اللَّه",
  "اللّٰه",
  "لله",
  "والله",
  "بالله",
  "تالله",
  "فالله",
]);

function isArabicLetter(ch: string): boolean {
  const o = ch.codePointAt(0) ?? 0;
  return (o >= 0x0621 && o <= 0x063a) || (o >= 0x0641 && o <= 0x064a) || "ةىآأإؤئءٱ".includes(ch);
}

export function stripHarakat(text: string): string {
  return [...text].filter((ch) => !HARAKAT.has(ch) && ch !== TATWEEL).join("");
}

const KEEP_RE = /[^\u0600-\u06FF\s*]+/g;
const MULTI_SPACE = /\s+/g;
const NUMBERING = /^[\d٠-٩]+[.\-)\]]\s*/;

export function splitBayt(text: string): string[] {
  if (!text || !text.trim()) return [];
  const t = text.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  let parts: string[];
  if (t.includes("***")) parts = t.split("***").map((p) => p.trim());
  else if (t.includes("\n")) parts = t.split("\n").map((p) => p.trim());
  else if (t.includes(" * ")) parts = t.split(" * ").map((p) => p.trim());
  else if ((t.match(/\*/g) || []).length === 1) parts = t.split("*").map((p) => p.trim());
  else parts = [t.trim()];
  return parts.filter(Boolean);
}

export function cleanLine(text: string): string {
  let t = text.trim().replaceAll(TATWEEL, "");
  t = t.replace(NUMBERING, "");
  t = t.replace(KEEP_RE, " ").replaceAll("*", " ");
  return t.replace(MULTI_SPACE, " ").trim();
}

type BaseLetter = { ch: string; sukun: boolean; vowel: boolean };

function iterBaseLetters(word: string): BaseLetter[] {
  const out: BaseLetter[] = [];
  let i = 0;
  while (i < word.length) {
    const ch = word[i]!;
    if (HARAKAT.has(ch) || ch === TATWEEL || ch === " ") {
      i += 1;
      continue;
    }
    if (!isArabicLetter(ch) && !"آأإؤئءىةٱ".includes(ch)) {
      i += 1;
      continue;
    }
    let j = i + 1;
    let hasShadda = false;
    let hasSukun = false;
    let hasVowel = false;
    while (j < word.length && HARAKAT.has(word[j]!)) {
      if (word[j] === SHADDA) hasShadda = true;
      else if (word[j] === SUKUN) hasSukun = true;
      else if (VOWEL_MARKS.has(word[j]!)) hasVowel = true;
      j += 1;
    }
    if (hasShadda) {
      out.push({ ch, sukun: true, vowel: false });
      out.push({ ch, sukun: false, vowel: true });
    } else {
      out.push({ ch, sukun: hasSukun, vowel: hasVowel });
    }
    i = j;
  }
  return out;
}

function isAllahWord(bare: string): boolean {
  return ALLAH_FORMS.has(bare) || bare.endsWith("الله") || bare === "الله";
}

function ph(
  char: string,
  kind: Kind,
  fixed: number | null,
  display: boolean,
  wordI: number,
  hint: string,
): Phoneme {
  return { char, kind, fixed, display, wordI, hint };
}

function allahPhonemes(
  word: string,
  wordI: number,
  firstWord: boolean,
  lastWord: boolean,
): Phoneme[] {
  let bare = stripHarakat(word);
  let prefix = "";
  for (const p of ["و", "ب", "ف", "ت", "ل"]) {
    if (bare.startsWith(p + "الله") || bare === p + "له") {
      prefix = p;
      bare = bare.slice(p.length);
      break;
    }
  }
  if (bare === "له") {
    prefix = prefix || "ل";
  }
  const out: Phoneme[] = [];
  if (prefix) out.push(ph(prefix, "cons", 1, true, wordI, "prefix"));
  if (firstWord && !prefix) {
    out.push(ph("ا", "cons", 1, true, wordI, "allah_alif"));
  } else {
    out.push(ph("ا", "wasl", null, true, wordI, "allah_alif"));
  }
  out.push(ph("ل", "cons", 0, true, wordI, "allah_lam1"));
  out.push(ph("ل", "cons", 1, true, wordI, "allah_lam2"));
  if (!lastWord) {
    out.push(ph("ا", "alif_madd", 0, false, wordI, "allah_madd"));
    out.push(ph("ه", "cons", null, true, wordI, "allah_ha"));
  } else {
    out.push(ph("ه", "cons", 0, true, wordI, "allah_ha"));
  }
  return out;
}

function function10(word: string, wordI: number): Phoneme[] {
  let letters = iterBaseLetters(word).map((x) => x.ch);
  if (!letters.length) letters = [...stripHarakat(word)];
  if (letters.length === 1) {
    const ch = letters[0]!;
    if (ALIF_MADD.has(ch) || ch === "ى") return [ph(ch, "alif_madd", 0, true, wordI, "fn")];
    return [ph(ch, "cons", 1, true, wordI, "fn"), ph("", "collapsed", 0, false, wordI, "fn_sukun")];
  }
  const out: Phoneme[] = [];
  letters.forEach((ch, k) => {
    if (k === 0) {
      let kind: Kind = "cons";
      if (ch === WAW) kind = "waw";
      else if (ch === YA || ch === "ى") kind = "ya";
      out.push(ph(ch, kind, 1, true, wordI, "fn10"));
    } else if (ALIF_MADD.has(ch) || ch === "ى" || ch === MADDA_ALIF) {
      out.push(ph(ch === MADDA_ALIF ? "ا" : ch, "alif_madd", 0, true, wordI, "fn10"));
    } else {
      out.push(ph(ch, "cons", 0, true, wordI, "fn10"));
    }
  });
  return out;
}

function pronoun(word: string, wordI: number, lastWord: boolean): Phoneme[] {
  let letters = iterBaseLetters(word).map((x) => x.ch);
  if (!letters.length) letters = [...stripHarakat(word)];
  const out: Phoneme[] = [];
  letters.forEach((ch, k) => {
    const last = lastWord && k === letters.length - 1;
    const fixed = last ? 0 : 1;
    out.push(ph(ch, "cons", fixed, true, wordI, "pron"));
  });
  return out;
}

function applyMaddConstraints(h: Hemistich) {
  const phs = h.phonemes;
  for (let i = 0; i < phs.length; i++) {
    const p = phs[i]!;
    if (p.kind === "alif_madd") {
      let j = i - 1;
      while (j >= 0 && (phs[j]!.kind === "wasl" || phs[j]!.kind === "collapsed")) j -= 1;
      if (j >= 0 && phs[j]!.kind !== "alif_madd") {
        if (phs[j]!.fixed === null) phs[j]!.fixed = 1;
      }
    }
    if (i > 0 && phs[i - 1]!.kind === "alif_madd" && p.fixed === 0) {
      if (p.hint === "allah_ha") continue;
      if (p.kind === "cons" || p.kind === "ta_marbuta" || p.kind === "collapsed") {
        p.kind = "collapsed";
      }
    }
  }
}

export function tokenizeHemistich(text: string): Hemistich {
  const cleaned = cleanLine(text);
  const words = cleaned.split(" ").filter(Boolean);
  const h: Hemistich = { raw: text, cleaned, words, phonemes: [] };
  if (!words.length) return h;

  for (let wi = 0; wi < words.length; wi++) {
    const word = words[wi]!;
    const bare = stripHarakat(word);
    const first = wi === 0;

    if (isAllahWord(bare)) {
      h.phonemes.push(...allahPhonemes(word, wi, first, wi === words.length - 1));
      continue;
    }
    if (FUNCTION_10.has(bare)) {
      h.phonemes.push(...function10(word, wi));
      continue;
    }
    if (PRONOUNS.has(bare)) {
      h.phonemes.push(...pronoun(word, wi, wi === words.length - 1));
      continue;
    }

    const letters = iterBaseLetters(word);
    let idx = 0;
    if (
      letters.length >= 3 &&
      (ALIF_MADD.has(letters[0]!.ch) || "أإٱآ".includes(letters[0]!.ch)) &&
      letters[1]!.ch === LAM
    ) {
      const shamsi = SHAMSI.has(letters[2]!.ch);
      const alifCh = letters[0]!.ch;
      if (alifCh === MADDA_ALIF) {
        h.phonemes.push(ph("آ", "cons", 1, true, wi, "hamza"));
        h.phonemes.push(ph("ا", "alif_madd", 0, true, wi, "madda"));
      } else if (first) {
        h.phonemes.push(ph("ا", "cons", 1, true, wi, "al_alif"));
      } else {
        h.phonemes.push(ph("ا", "wasl", null, true, wi, "al_wasl"));
      }
      if (shamsi) {
        const sham = letters[2]!.ch;
        h.phonemes.push(ph(sham, "cons", 0, true, wi, "shamsi_sukun"));
        h.phonemes.push(ph(sham, "cons", 1, false, wi, "shamsi_move"));
        idx = 3;
      } else {
        h.phonemes.push(ph("ل", "cons", 0, true, wi, "qamari_lam"));
        idx = 2;
      }
    }

    const rest = letters.slice(idx);
    rest.forEach((item, k) => {
      const ch = item.ch;
      const lastOfWord = k === rest.length - 1;
      const lastOfLine = lastOfWord && wi === words.length - 1;
      if (ch === MADDA_ALIF) {
        h.phonemes.push(ph("أ", "cons", 1, true, wi, "madda_h"));
        h.phonemes.push(ph("ا", "alif_madd", 0, true, wi, "madda_a"));
        return;
      }
      if (ALIF_MADD.has(ch)) {
        h.phonemes.push(ph("ا", "alif_madd", 0, true, wi, "madd"));
        return;
      }
      if (ch === "ى") {
        h.phonemes.push(ph("ى", "alif_madd", 0, true, wi, "maqsur"));
        return;
      }
      if (ch === TA_MARBUTA) {
        h.phonemes.push(ph("ة", "ta_marbuta", lastOfLine ? 0 : null, true, wi, "ta"));
        return;
      }
      let kind: Kind = "cons";
      if (ch === WAW) kind = "waw";
      else if (ch === YA) kind = "ya";
      let fixed: number | null = null;
      if (item.sukun) fixed = 0;
      else if (item.vowel) fixed = 1;
      if (lastOfLine) fixed = 0;
      if (kind === "waw" || kind === "ya") {
        if (!lastOfLine && !first && k > 0) {
          if (item.vowel) fixed = 1;
          else if (item.sukun) fixed = 0;
          else fixed = null;
        }
      }
      if (k === 0 && idx === 0 && rest.length === 1 && PREFIX_CONNECT.has(ch) && !lastOfLine) {
        fixed = 1;
      }
      h.phonemes.push(ph(ch, kind, fixed, true, wi, kind));
    });
  }

  if (h.phonemes.length) {
    for (let i = h.phonemes.length - 1; i >= 0; i--) {
      const p = h.phonemes[i]!;
      if (p.display || p.kind !== "collapsed") {
        if (p.kind !== "wasl") p.fixed = 0;
        break;
      }
    }
  }
  applyMaddConstraints(h);
  return h;
}
